// binary-object/dist/msgpack.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MsgpackSource = exports.MsgpackSink = void 0;
const msgpack_1 = require("msgpack");
const binary_object_1 = require("./binary-object");
const index_1 = require("./index");
class MsgpackSink extends index_1.Sink {
    constructor(sink) {
        super();
        this.sink = sink;
    }
    write(data) {
        const buffer = msgpack_1.pack(data);
        binary_object_1.encodeNumber(this.sink, buffer.byteLength);
        this.sink.writeBuffer(buffer, 0, buffer.byteLength);
    }
    close() {
        this.sink.write(binary_object_1.Types.End);
        this.sink.close();
    }
}
exports.MsgpackSink = MsgpackSink;
class MsgpackSource extends index_1.Source {
    constructor(source) {
        super();
        this.source = source;
    }
    read() {
        const byteLength = binary_object_1.decodeNumber(this.source);
        const buffer = this.source.readBatch(byteLength).slice(0, byteLength);
        return msgpack_1.unpack(buffer);
    }
    close() {
        this.source.close();
    }
    *iterator(options) {
        for (;;) {
            const byteLength = binary_object_1.decodeNumberOrEnd(this.source);
            if (byteLength === index_1.End) {
                if (options === null || options === void 0 ? void 0 : options.autoClose) {
                    this.source.close();
                }
                break;
            }
            const buffer = this.source.readBatch(byteLength).slice(0, byteLength);
            yield msgpack_1.unpack(buffer);
        }
    }
}
exports.MsgpackSource = MsgpackSource;
